spec/edpm-openstack-service-architecture.md

```diff
@@ -0,0 +1,326 @@
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneNode
+metadata:
+    name: edpm-compute-0
+spec:
+    ansibleHost: 192.168.122.100
+    hostName: edpm-compute-0
+    node:
+        ansibleSSHPrivateKeySecret: dataplane-ansible-ssh-private-key-secret
+    role: edpm-compute
+...
+```
+The `novaTemplate` is removed from the `node`. As this DataPlaneNode refers
+to a DataPlaneRole that has a `nova.openstack.org/cell` annotation, this
+DataPlaneNode is a compute node connected to cell1.
+
+OpenstackDataPlaneNode/edpm-compute-1
+```yaml
+apiVersion: dataplane.openstack.org/v1beta1
+kind: OpenStackDataPlaneNode
+metadata:
+    name: edpm-compute-1
+    annotations:
+        nova.openstack.org/cell: nova-cell2
+spec:
+    ansibleHost: 192.168.122.101
+    hostName: edpm-compute-1
+    node:
+        ansibleSSHPrivateKeySecret: dataplane-ansible-ssh-private-key-secret
+    role: edpm-compute
+...
+```
+This DataPlaneNode has an `nova.openstack.org/cell` annotation that overrides
+the cell annotation of its role. So this EDPM node is a compute node connected
+to cell2.
+
+NovaExternalCompute=edpm-compute-0
+```yaml
+apiVersion: nova.openstack.org/v1beta1
+kind: NovaExternalCompute
+metadata:
+    name: edpm-compute-0
+spec:
+    dataplaneNodeName: edpm-compute-0
+    cellName: nova-cell1
+    customServiceConfig: |
+    # human deployer can set this to pass through service specific config
+    sshKeySecretName: dataplane-ansible-ssh-private-key-secret
+...
+```
+This CR is auto created by the nova-operator based on the corresponding
+DataPlaneNode CR and the cell annotation set on it. The spec fields, except
+`CustomServiceConfig` is populated by the nova-operator. The
+`CustomServiceConfig` field is never set by the nova-operator it can be used
+by the human deployer.
+
+The following fields will be copied over from the DataplaneNode (or Role due to
+inheritance) and changes there will be propagated to the NovaExternalCompute CR
+* InventoryConfigMapName
+* NetworkAttachments
+* SshKeySecretName
+
+The AnsibleEEContainerImage, NovaComputeContainerImage, and
+NovaLibvirtContainerImage will be set based on the respective ENV value via the
+defaulting webhook in nova-operator during NovaExternalCompute auto creation
+but these fields can be changed later independently from their defaults.
+
+The NovaInstance field is removed and the CellName field is redefined to hold
+the name of the NovaCell CR. The name of the NovaCell CR is a combination of
+the name of the cell from the NovaCellTemplate in the Nova CR and the name of
+the Nova CR. So it uniquely identifies a cell even if there are multiple Nova
+CRs exist in the namespace.
+
+The upstream nova project does not support moving computes between cells.
+Therefore the NovaCell field can only be set at creation and cannot be updated
+later. The dataplane-operator needs to implement a validation hook to prevent
+the related annotation to change after being set.
```

Comment:
hum so it could but i dont think it must do this.

instead we could have each service operator install a validating webhook for the dataplanenode/role CR that would just validate there annotation nothing else.

that way the dataplonenode does not need to have any knowledge of what the annotation are or do.